src/models/WidgetTemplateImport.php

```diff
@@ -0,0 +1,212 @@
+                    $this->addError('tarFiles', \Yii::t('widgets', 'Error while saving file'));
+                }
+            }
+            return true;
+        }
+
+        return false;
+    }
+
+    public function attributeLabels()
+    {
+        $attributeLabels = parent::attributeLabels();
+        $attributeLabels['tarFiles'] = \Yii::t('widgets', 'Files');
+        return $attributeLabels;
+    }
+
+    public function removeImportDirectory(): bool
+    {
+        try {
+            FileHelper::removeDirectory($this->importDirectory);
+            return true;
+        } catch (\ErrorException $e) {
+            \Yii::error($e->getMessage());
+            return false;
+        }
+    }
+
+    protected function extractTar(string $filepath, string $extractDirectory): bool
+    {
+        return (new \PharData($filepath))->extractTo($extractDirectory);
+    }
+
+    public function extractFiles(): bool
+    {
+        foreach ($this->tarFiles as $file) {
+            $filepath = $this->importDirectory . DIRECTORY_SEPARATOR . $file->baseName . '.' . $file->extension;
+            if ($this->extractTar($filepath, $this->extractDirectory($file->baseName)) === false) {
+                $this->addError('tarFiles', \Yii::t('widgets', 'Error while extracting file'));
+                return false;
+            }
+        }
+        return true;
+    }
+
+    protected function extractDirectory(string $filename): string
+    {
+        return $this->importDirectory . DIRECTORY_SEPARATOR . Inflector::slug(basename($filename));
+    }
+
+    public function import()
+    {
+        $transaction = WidgetTemplate::getDb()->beginTransaction();
+        if ($transaction && $transaction->getIsActive()) {
+            foreach ($this->tarFiles as $file) {
+                $extractDirectory = $this->extractDirectory($file->baseName);
+                if ($this->importTemplateByDirectory($extractDirectory) === false) {
+                    $this->addError('tarFiles', \Yii::t('widgets', 'Error while importing widget template'));
+                    $transaction->rollBack();
+                    return false;
+                }
+
+                FileHelper::removeDirectory($extractDirectory);
+            }
+            $transaction->commit();
+            return true;
+        }
+        return false;
+    }
+
+    protected function importTemplateByDirectory(string $extractDirectory): bool
+    {
+        $meta = $this->templateMeta($extractDirectory);
+        $template = $this->templateContent($extractDirectory);
+        $schema = $this->schemaContent($extractDirectory);
+
+        $model = new WidgetTemplate([
+            'name' => $meta['name'],
+            'php_class' => $meta['php_class'],
+            'twig_template' => $template,
+            'json_schema' => $schema
+        ]);
+
+        if ($model->save() === false) {
+            \Yii::error($model->getErrors());
```

Comment:
These errors should be passed to the user so that he gets information why the import did not work. (Exceptions?)
see: Line 136

Reply:
I've added a custom exceptions. This bundles the error messages that occur when saving. The exception is then caught in the controller and passed on to the user.